File: crashreporter/src/main/java/net/cebular/crashreporter/ui/ReportListActivity.java
package net.cebular.crashreporter.ui;

import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.os.Bundle;
import com.google.android.material.tabs.TabLayout;
import androidx.viewpager.widget.ViewPager;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;

import net.cebular.crashreporter.R;
import net.cebular.crashreporter.adapter.MainPagerAdapter;
import net.cebular.crashreporter.utils.Constants;

public class ReportListActivity extends AppCompatActivity {

    private MainPagerAdapter mainPagerAdapter;
    private int selectedTabPosition = 0;

    //

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_report_list);

        Toolbar toolbar = findViewById(R.id.toolbar);
        toolbar.setTitle(getString(R.string.crash_reporter));
        toolbar.setSubtitle("App: " + getApplicationName());
        setSupportActionBar(toolbar);

        ViewPager viewPager = findViewById(R.id.viewPager);
        setupViewPager(viewPager);

        TabLayout tabLayout = findViewById(R.id.tabLayout);
        tabLayout.setupWithViewPager(viewPager);
    }

    //

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.activity_report_list_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // menu actions are delegated to ReportsFragment
        return mainPagerAdapter.getItem(selectedTabPosition).onOptionsItemSelected(item);
    }

    private void setupViewPager(ViewPager viewPager) {
        mainPagerAdapter = new MainPagerAdapter(getSupportFragmentManager());
        viewPager.setAdapter(mainPagerAdapter);

        viewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageSelected(int position) {
                selectedTabPosition = position;
            }
            @Override public void onPageScrolled(int i, float v, int i1) { }
            @Override public void onPageScrollStateChanged(int i) { }
        });

        Intent intent = getIntent();
        if (intent != null && intent.hasExtra(Constants.EXTRA_LANDING)) {
            int reportType = intent.getIntExtra(Constants.EXTRA_LANDING, 0);
            if (reportType == Constants.REPORT_TYPE_EXCEPTION) {
                selectedTabPosition = 1;
            } else {
                selectedTabPosition = 0;
            }
        }
        viewPager.setCurrentItem(selectedTabPosition);
    }

    private String getApplicationName() {
        ApplicationInfo applicationInfo = getApplicationInfo();
        int stringId = applicationInfo.labelRes;
        return stringId == 0 ? applicationInfo.nonLocalizedLabel.toString() : getString(stringId);
    }

}

File: crashreporter/src/main/java/net/cebular/crashreporter/internal/ReportListActivity.kt
package net.cebular.crashreporter.internal

import android.content.Context
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import net.cebular.crashreporter.R
import net.cebular.crashreporter.databinding.ActivityReportListBinding

class ReportListActivity : AppCompatActivity() {

    private lateinit var binding: ActivityReportListBinding
    private lateinit var reportsAdapter: ReportsAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityReportListBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setSupportActionBar(binding.toolbar)
        binding.toolbar.title = getString(R.string.crash_reporter)
        binding.toolbar.subtitle = getApplicationName()

        reportsAdapter = ReportsAdapter { reportFile ->
            ReportDetailActivity.launchActivity(this@ReportListActivity, reportFile.name)
        }

        binding.recyclerViewReports.apply {
            layoutManager = LinearLayoutManager(this@ReportListActivity, LinearLayoutManager.VERTICAL, false)
            addItemDecoration(DividerItemDecoration(this@ReportListActivity, LinearLayoutManager.VERTICAL))
            adapter = reportsAdapter
        }
    }

    override fun onResume() {
        super.onResume()
        loadReports()
    }

    //

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.activity_report_list_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when(item.itemId) {
            R.id.action_clear_logs -> {
                AlertDialog.Builder(this)
                    .setTitle("Clear logs?")
                    .setMessage("This will permanently remove all logs.")
                    .setPositiveButton(android.R.string.ok) { dialog, _ ->
                        clearReports()
                    }
                    .setNegativeButton(android.R.string.cancel) { dialog, _ ->
                        // Do nothing.
                    }
                    .show()
                true
            }

            else -> super.onOptionsItemSelected(item)
        }
    }

    //

    private fun loadReports() {
        getCrashReportsDir()?.listFiles()?.let {
            reportsAdapter.setReports(it.sortedByDescending { f -> f.name })
        }
    }

    private fun clearReports() {
        getCrashReportsDir()?.listFiles()?.forEach {
            it.delete()
        }
        loadReports()
    }

    private fun Context.getApplicationName(): String {
        return applicationInfo.let {
            if (it.labelRes == 0) {
                it.nonLocalizedLabel.toString()
            } else {
                getString(it.labelRes)
            }
        }
    }

}